flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -0,0 +1,246 @@
+
+    @throws(classOf[Exception])
+    override def open(configuration: Configuration): Unit = {
+      val list = this.getRuntimeContext.
+        getBroadcastVariable[WeightVector](WEIGHTVECTOR_BROADCAST)
+
+      weightVector = list.get(0)
+    }
+
+    override def map(example: LabeledVector): (WeightVector, Double, Int) = {
+
+      val lossFunction = parameterMap(LossFunction)
+      val regType = parameterMap(RegularizationType)
+      val regParameter = parameterMap(RegularizationParameter)
+      val predictionFunction = parameterMap(PredictionFunction)
+      val dimensions = example.vector.size
+      // TODO(tvas): Any point in carrying the weightGradient vector for in-place replacement?
+      // The idea in spark is to avoid object creation, but here we have to do it anyway
+      val weightGradient = new DenseVector(new Array[Double](dimensions))
+
+
+      // TODO(tvas): Indentation here?
+      val (loss, lossDeriv) = lossFunction.lossAndGradient(
+                                example,
+                                weightVector,
+                                weightGradient,
+                                regType,
+                                regParameter,
+                                predictionFunction)
+
+      // Restrict the value of the loss derivative to avoid numerical instabilities
+      // TODO(tvas): Do this for all elements of the weighGradient as well?
+      val restrictedLossDeriv: Double = {
+        if (lossDeriv < -IterativeSolver.MAX_DLOSS) {
+          -IterativeSolver.MAX_DLOSS
+        }
+        else if (lossDeriv > IterativeSolver.MAX_DLOSS) {
+          IterativeSolver.MAX_DLOSS
+        }
+        else {
+          lossDeriv
+        }
+      }
+      // TODO(tvas): Any way to avoid object creation here?
+      (new WeightVector(weightGradient, restrictedLossDeriv), loss, 1)
+    }
+  }
+
+  /** Performs the update of the weights, according to the given gradients and regularization type.
+    *
+    */
+  private class WeightsUpdate() extends
+  RichMapFunction[(WeightVector, Double, Int), WeightVector] {
+
+    var weightVector: WeightVector = null
+
+    @throws(classOf[Exception])
+    override def open(configuration: Configuration): Unit = {
+      val list = this.getRuntimeContext.
+        getBroadcastVariable[WeightVector](WEIGHTVECTOR_BROADCAST)
+
+      weightVector = list.get(0)
+    }
+
+    override def map(gradientLossAndCount: (WeightVector, Double, Int)): WeightVector = {
+      val regType = parameterMap(RegularizationType)
+      val regParameter = parameterMap(RegularizationParameter)
+      val stepsize = parameterMap(Stepsize)
+      val weightGradients = gradientLossAndCount._1
+      val lossSum = gradientLossAndCount._2
+      val count = gradientLossAndCount._3
+
+      // Scale the gradients according to batch size
+      BLAS.scal(1.0/count, weightGradients.weights)
+
+      // Add the regularization term to the gradient, in-place
+      val adjustedLoss = {
```

Comment:
Why is the loss only adjusted if the regularization function is differentiable? Shouldn't that be independent of each other?